src/main/java/com/zendesk/maxwell/producer/MaxwellKinesisProducer.java

```diff
@@ -32,82 +31,168 @@
 	private final String json;
 	private final String key;
 
-	public KinesisCallback(AbstractAsyncProducer.CallbackCompleter cc, BinlogPosition position, String key, String json) {
+	private AtomicInteger successRecords;
+	private AtomicInteger errorRecords;
+	private final KinesisProducer kinesisProducer;
+	private final String kinesisStream;
+	private int attempts = 0;
+	private final int maxAttempts;
+
+	public KinesisCallback(AbstractAsyncProducer.CallbackCompleter cc, BinlogPosition position, String key, String json, AtomicInteger successRecords, AtomicInteger errorRecords, KinesisProducer kinesisProducer, String kinesisStream, int attempts, int maxAttempts) {
 		this.cc = cc;
 		this.position = position;
 		this.key = key;
 		this.json = json;
+		this.successRecords = successRecords;
+		this.errorRecords = errorRecords;
+		this.kinesisProducer = kinesisProducer;
+		this.kinesisStream = kinesisStream;
+		this.attempts = attempts;
+		this.maxAttempts = maxAttempts;
 	}
 
 	@Override
 	public void onFailure(Throwable t) {
 		logger.error(t.getClass().getSimpleName() + " @ " + position + " -- " + key);
 		logger.error(t.getLocalizedMessage());
 
-		if(t instanceof UserRecordFailedException) {
-			Attempt last = Iterables.getLast(((UserRecordFailedException) t).getResult().getAttempts());
-			logger.error(String.format("Record failed to put - %s : %s", last.getErrorCode(), last.getErrorMessage()));
-		}
+		if(attempts>=maxAttempts) {
+			int errRecords = errorRecords.incrementAndGet();
+
+			if(t instanceof UserRecordFailedException) {
+				Attempt last = Iterables.getLast(((UserRecordFailedException) t).getResult().getAttempts());
+				UserRecordResult result = ((UserRecordFailedException) t).getResult();
+				List<String> errorList = new ArrayList<String>();
+				for(Attempt attempt : result.getAttempts()){
+				   errorList.add(String.format(
+						   "Delay after prev attempt: %d ms, "
+								   + "Duration: %d ms, Code: %s, "
+								   + "Message: %s",
+						   attempt.getDelay(), attempt.getDuration(),
+						   attempt.getErrorCode(),
+						   attempt.getErrorMessage()));
+				}
+				String errorListStr = StringUtils.join(errorList, "n");
+				logger.error(String.format(
+						"Record failed to put, attempts: %d. Errors: %s", result.getAttempts().size(), errorListStr));
+				logger.error(String.format("Number of failed records up to now: %d", errRecords));
+				logger.error(String.format("Errored message: %s", json));
+			}
+
+			cc.markCompleted();
+		} else {
+			attempts++;
 
-		logger.error("Exception during put", t);
+			logger.warn(String.format("Record failed to put. Retrying. Attempt: %d ", attempts));
 
-		cc.markCompleted();
+			try {
+				ByteBuffer encodedValue = ByteBuffer.wrap(json.getBytes("UTF-8"));
```

Comment:
Any way to DRY this up? The code here is very similar to the code in the main handler.